x-pack/solutions/observability/plugins/apm/public/components/shared/trace_waterfall/use_trace_waterfall.ts

```diff
@@ -95,15 +106,79 @@ export function getTraceParentChildrenMap(traceItems: TraceItem[]) {
   return traceMap;
 }
 
-export function getTraceWaterfall(
+export enum TraceDataState {
+  Full = 'full',
+  Partial = 'partial',
+  Empty = 'empty',
+}
+
+export function getRootItemOrFallback(
+  traceParentChildrenMap: Record<string, TraceItem[]>,
+  traceItems: TraceItem[]
+) {
+  if (traceItems.length === 0) {
+    return {
+      traceState: TraceDataState.Empty,
+    };
+  }
+
+  const rootItem = traceParentChildrenMap.root?.[0];
+
+  const parentIds = new Set(traceItems.map(({ id }) => id));
+  const orphans = traceItems.filter((item) => item.parentId && !parentIds.has(item.parentId));
+
+  if (rootItem) {
+    return {
+      traceState: orphans.length === 0 ? TraceDataState.Full : TraceDataState.Partial,
+      rootItem,
+      orphans,
+    };
+  }
+
+  const [fallbackRootItem, ...remainingOrphans] = orphans;
```

Comment:
q: What would be the `fallbackRootItem` here? A `span`, `transaction` or doesn't matter?

Reply:
>What would be the fallbackRootItem here? A span, transaction or doesn't matter?

@crespocarlos for the new unified trace waterfall we removed the span/transaction context. The waterfall doesn't know about that, it only know its own interface.